Add CLI batch support for presenting warnings when pasting configs

etracer65 opened the pull request:
Adds a new `batch` command used to delineate a group of commands as a related batch. The primary purpose is to group commands sent from the Configurator resulting from a copy/pasted config. Currently the output of the `diff all` command appends a `save` at the end of its output. So when a user pastes in their config to restore they may not see any errors because the `save` command causes the flight controller to reboot before they can review. When commands are wrapped inside of a batch, any errors will set a flag that will issue a warning when the `save` command is executed. This allows the user ro review and correct their configuration.

While the `batch` command can be used interactively, it's really meant for the Configurator to send when needed for multi-line command pasting.  As such, see related PR for the Configurator: https://github.com/betaflight/betaflight-configurator/pull/1193

Example: Assume a command batch like:
```
batch start
defaults nosave
set p_pitch = 48
set i_pitch = 55
set rc_smoothing_type = OFF
set rc_interp_ch = RPYT
save
```

In this example the `batch end` is redundant because the commands contain a save.

Since the batch contains an error the `save` will be deferred:
```
# batch start
Command batch started

# defaults nosave

# resetting to defaults

# set p_pitch = 48
p_pitch set to 48
# set i_pitch = 55
i_pitch set to 55
# set rc_smoothing_type = OFF
###ERROR### Invalid value
Allowed values: INTERPOLATION, FILTER

# set rc_interp_ch = RPYT
rc_interp_ch set to RPYT
# save
###WARNING### Errors were detected - please review before continuing
###WARNING### Enter the 'save' command again to continue
```
The user can now review and fix the problem. Alternately they can just issue another `save` and it will not be prevented.

DieHertz commented on src/main/interface/cli.c:
```diff
@@ -3239,10 +3262,54 @@ static void cliDumpRateProfile(uint8_t rateProfileIndex, uint8_t dumpMask)
+
+static void cliBatch(char *cmdline)
+{
+    if (strncasecmp(cmdline, "start", 5) == 0) {
```
I don't like N component as it will allow substring match, not whole match.
E.g. `batch starting` will also pass the check

mikeller replied:
I don't think this is needed at all:
- errors that are made even without the user (or some app by means of hidden magic) explicitly enabling an 'error detect' mode do still matter;
- users can still save, even if there were errors, so there is little value to suppressing error detection for any commands entered in CLI.

If this stays, it should be renamed to `errordetect` or suchlike, since the functionality per se has got nothing to do with batching.

mikeller commented on src/main/interface/cli.c:
```diff
@@ -3239,10 +3262,54 @@ static void cliDumpRateProfile(uint8_t rateProfileIndex, uint8_t dumpMask)
 
+#ifdef USE_CLI_BATCH
+    if (commandBatchActive && commandBatchError) {
+        cliPrintCommandBatchWarning("Enter the 'save' command again to continue");
```
I don't like that there is state introduced when none is needed - the command to force save even in the presence of errors should be `save force` or suchlike, not just a repeated `save`.

ledvinap commented:
Another approach is to implement error (and possibly warning) counters + command to display and reset counters.
The dump may simply reset errors at beginning  and display errors before save.
resetting errors or `save force` will be necessary to commit changes ...

ledvinap commented:
@etracer65 : 

- counting errors will make it easier to review output

- `save` can simply fail if errors are encountered between `batch start` and `batch end` (without resetting error counted, with minimal error message. `batch end` can instruct user)

- `batch end` can be used always (even after `save`) and print warning/error encountered

etracer65 commented:
@ledvinap 

> counting errors will make it easier to review output

I'm not opposed to that.  Displaying something like "3 errors were detected, please review before continuing".  However I think the error count should be relative to each batch and reset on the `batch start`.  That way the reporting is always relative to what the user just did.

> save can simply fail if errors are encountered between batch start and batch end (without resetting error counted, with minimal error message. batch end can instruct user)

That's what is already happening. A `save` within the batch will just turn into a warning if errors were detected in the batch prior to the save.  Of course if there were no error the `save` completes normally and the batch is terminated since the FC reboots.

> batch end can be used always (even after save) and print warning/error encountered

The only reason I suppressed the `batch end` if the commands contained a `save` was for cosmetic reasons.  In the case of the `save` reporting warnings about errors I wanted that to be the last thing displayed and most evident to the user.  But it's not a big deal to me one way or the other.

etracer65 commented:
Also note that this proposed solution with batch support and enhancements to the Configurator to automatically wrap multi-line command blocks won't break any other scripting solutions.  Solutions that connect to the flight controller to configure it won't be aware of the `batch` commands and as such they'll behave just as they do now.  Of course they'd be free to make them batch-aware if they wanted - possibly making them more robust.

On the other hand implementing some solution that tries to enforce a good configuration by requiring some kind of `save force` method after errors are noticed will break scripting solutions.

ledvinap commented:
@etracer65 : The idea is to decouple `save` from `batch` logic - `save` will always fail inside batch block with error, `batch end` will terminate batch and allow `save` to work normally ...
Otherwise I agree with your reasoning ...

etracer65 commented:
@ledvinap I get what you're saying. But note that currently a `save` inside of a batch implicitly terminates the batch just as if a `batch end` was processed.  So a subsequent save will already function normally.  Like I said, it's really just a cosmetic thing to not have the `batch end` show up in the output when not needed.

ledvinap commented:
@etracer65 : Well, I don't like the 'implicit termination' part. From programmers viewpoint it introduces unnecessary dependency ...

etracer65 commented:
@ledvinap Well actually I guess it's more of an explicit batch termination because commands after a `save` really aren't valid anyway as they'll never get interpreted.  Same applies to `defaults`, `bl`, `exit`, and 'msc' as they all force a reboot.  But none of these others need special handling (suppression) as they don't update settings.

mikeller commented:
@etracer65: I really do not like that you are proposing to create a 'two class CLI society', and separate between those who blindly copy / paste stuff into the configurator, and rely on some hidden magic to do things for them, and not doing anything in all other cases. I don't see what should be wrong with, if we go through the effort to add validation, doing this validation in all cases, and not just when configurator sprinkles in some hidden commands that the users will be left wondering where they came from, and why this will not work if they don't use the configurator.

etracer65 commented:
@mikeller My goals:

- address the 99.9% case where people are accessing the CLI using the Configurator.
- solve the problem of pasted configs having errors that unnoticed.
- not interfere with interactive "one at a time" commands.
- not break other mechanisms that use CLI (outside the configurator) for scripting.

So what I've implemented so far seamlessly accomplishes these goals.  If there are better ways to accomplish then I'm open to alternatives.  I understand your point about differentiating between Configurator CLI and "other" CLI users but we also have to consider the tiny minority involved.  And we're not breaking or making anything worse for them but at the same time improving for the vast majority.  Also I'm not sure how to reconcile the last point about about not breaking external scripting that might use the CLI outside the configurator.

etracer65 commented on src/main/interface/cli.c:
```diff
@@ -374,16 +379,34 @@ static void cliPrintLinef(const char *format, ...)
+#ifdef USE_CLI_BATCH
+static void cliPrintWarningLinef(const char *format, ...)
+{
+    cliPrint("###WARNING### ");
```
It's following the same consistent style with the `###ERROR###` tag and the `###CORRUPTED CONFIG###` tag  (#6697).  Also have plans to use them as markers to identify errors/warnings in the CLI output and colorize them for better visibility.

mikeller commented:
@etracer65: I think we want the same in the end. I just think that introducing extra syntax in the firmware, and then patching configurator to hide away the extra syntax from users, is not the best way to do this:
- it will confuse users by adding unrequested things to the CLI input;
- it will not have the desired effect for anything that is not using a configurator version with this patch (most likely more than 1% of users);
- the double save is unintuitive, and will confuse users as well - better use an explicit modifier to force save.

In addition to this I do not think that the approach to 'not break other uses of the CLI' is the right thing to do: What we are doing is a new feature (entry validation), it can be overridden by a 'save force` if desired, and we should roll it out to all uses of the CLI. If anything is currently trying to set invalid values it is broken as it is, and not likely to work in the way it is intended to, so it will need fixing whether we force it to supply valid values or not.

mikeller commented:
@etracer65:

> I think this will just lead to people using save force all the time negating the whole improvement.

If this is the case then we are in exactly the place where we are anyway with what you are proposing, with no protection whatsoever for anything that is not copy pasted into the configurator, so I can't see how this is an argument against doing implicit validation.

etracer65 commented:
How is that the case?  Maybe you don't fully understand my proposal.  There is no `save force`.

In interactive mode nothing is changed and the behavior is exactly as it is now.  The user makes a mistake, sees the error, fixes it, then saves.  The save command is not interfered with at all. There's no need because the user can see the errors as they make them and are aware of the problem.  Nothing to fix while in interactive "command at a time" use case.

In batch mode we're trying to fix the errors scrolling off the screen or a `save` rebooting the flight controller before the user sees the errors.  The `batch start` clues the firmware that a group of commands are coming and if any  errors occur then don't execute the save (with a warning).  Like in interactive mode the user now gets a chance to see the errors and resolve them.  **The only `save` that's deferred is the one that's included in the batch (but only if there are errors)**.  So any interactively issued saves work without any interruption.  If the batch didn't contain a save but had errors then warnings are still generated to notify the user because the errors might have scrolled off.  The hope is that the user fixes the errors with some interactive commands, but either way they're free to issue an interactive `save` like always.

mikeller commented:
Yes I do understand you. If users use `save force` for their hand entered content then there is no validation whatsoever for hand entered settings, just like there isn't in your proposal. Except that in my proposal this is the worst case, in yours it is default.
Making the function of a command change based on repetition is a bad idea on my opinion, it will make adoption of input validation by other systems that use CLI hard to impossible, and should be avoided at all cost.

etracer65 commented:
> If users use save force for their hand entered content then there is no validation whatsoever for hand entered settings, just like there isn't in your proposal. Except that in my proposal this is the worst case, in yours it is default.

Of course there's validation.  After each command the user gets feedback on the result of the command. Just like there is now.  The only difference is that in your proposal the user is going to be annoyed by having the `save` fail (even though they fixed the problem) and then forced to save again with `save force`.  So they'll just get into the habit of using `save force` all the time to avoid frustration and the any value the `save` prevention provides will be nullified.

> Making the function of a command change based on repetition is a bad idea on my opinion

The function of the command doesn't change - it's purpose is to `save` and nothing else. The change is that if the command is embedded in a batch with errors then it will reject until the warning is displayed for the batch.  In your proposal the `save` command rejects if there's any error in the entire session with no recourse for the user other than to use another command to save.

>  it will make adoption of input validation by other systems that use CLI hard to impossible, and should be avoided at all cost.

How so? I can't really see any change whatsoever to non-Configurator CLI use.  Any kind of scripted type of CLI should be looking at the result of each command already anyway.  We shouldn't be encouraging a method that blindly sends commands and hopes to catch the errors on the `save`.  Furthermore if other CLI users/scripting want to they're free to adopt the improved batching behavior by utilizing `batch start` and `batch end` - but they're not required to.

On the other hand your proposal will break any existing CLI scripting as the `save` command will behave unexpectedly.  In which case they'll switch to `save force` to return to the previous behavior.

mikeller commented:
@etracer65:

> Of course there's validation. After each command the user gets feedback on the result of the command. Just like there is now. The only difference is that in your 

Ok, if there is validation, as you say, then we don't need to change anything. But you seem to think that we need to add more validation. I just disagree that we should do this in a 'two class society' way, where only users of the CLI in configurator can benefit of this extra validation without adding their own hidden magic and extra complexity.

> proposal the user is going to be annoyed by having the save fail (even though they fixed the problem) and then forced to save again with save force. So they'll just get into the habit of using save force all the time to avoid frustration and the any value the save prevention provides will be nullified.

What makes you think that the users will not just get into the habit of blindly typing `save` until the save succeeds if you are assuming that they will start to blindly type `save force`?

The function of the command doesn't change - it's purpose is to save and nothing else. The change is that if the command is embedded in a batch with errors then it 

Of course it changes - how can you say that 'save only if there were no input errors' and 'force save even in the presence of invalid input' is the same?

> it will make adoption of input validation by other systems that use CLI hard to impossible, and should be avoided at all cost.

> How so? I can't really see any change whatsoever to non-Configurator CLI use. Any kind of scripted type of CLI should be looking at the result of each command already anyway. We shouldn't be encouraging a method that blindly sends commands and hopes to catch the errors on the save. Furthermore if other CLI users/scripting 

Again, if this is your opinion, then why are you proposing this anyway. We should either have extra validation of input like you propose, or not have it, but not use a mix that relies on decorating input with extra syntax, and hidden magic in just one application to hide the extra syntax requirements away from users.

Also, if this should ever go ahead, we really have to stop calling this anything like `batch`, since it has nothing whatsoever to do with batching, but with input validation.

But the longer I think about it, the less I can see why adding this extra safety net for all input would be a bad thing - even if users copy / paste single lines from GitHub or RCgroups, or are trying out a command with an invalid range. Telling them that something went wrong, and asking them to confirm that the settings are the way they want isn't wrong in my opinion.

etracer65 commented:
Ok I revived and rebased this. Added `batch start` to the top of the `diff` and `dump` outputs. If the output produces a `save` like `diff all` or `dump all` then the additional `batch end` is omitted as it's unnecessary. Otherwise the `batch end` is appended at the bottom of the output.

So if a user takes the outputs and later copy/pastes them into the configurator, then any detected errors will present a warning to the user and prevent the initial save (if applicable). The user can then review and fix the problems and at any point issue another save.

I left the `#define USE_CLI_BATCH` where it was which basically means this is enabled for every target for now. I guess we'll see how we do with target overflows and decide from there.

mikeller commented on src/main/cli/cli.c:
```diff
@@ -3788,6 +3843,9 @@ static void cliDefaults(char *cmdline)
 
     if (saveConfigs) {
+#ifdef USE_CLI_BATCH
+        resetCommandBatch();
```
This should probably be outside of the `if` - if the config is wiped with a `defaults nosave` we effectively get a clean slate, and there wont be the potential for previous failed commands affecting us.

etracer65 replied:
Yes, that makes sense - I'll change.

etracer65 replied:
The downside is that it makes the ordering of the `batch start` command more sensitive. If it's not after the `defaults nosave` it won't catch any errors. But it is correctly positioned in the `diff` and `dump` outputs.

mikeller replied:
I guess we could have the `defaults` reset just any errors that were found, but not change the on/off status for the batch mode.

etracer65 replied:
Yes, this will work. But it still kind of defeats the underlying goal but I'm not sure there's a better fix. The idea was to tell them of any errors in the block they pasted that might have scrolled off or otherwise obscured by the reboot. So logically commands that fail before the `defaults nosave` while not having any real effect, are technically still errors in the config they tried to use. But their config was wrong to begin with anyway and would never have worked so I'm thinking we can only go so far.

etracer65 replied:
Pushed a commit to only reset the error state on `defaults nosave`.

mikeller commented on src/main/cli/cli.c:
```diff
@@ -3745,10 +3756,54 @@ static void cliDumpRateProfile(uint8_t rateProfileIndex, uint8_t dumpMask)
     UNUSED(cmdline);
 
+#ifdef USE_CLI_BATCH
+    if (commandBatchActive && commandBatchError) {
```
I still don't like that this makes the `save` command context sensitive - it would be much cleaner if a `save force` or similar was required to persist if there were errors, not just a repetition of the same command until the board resets.

etracer65 replied:
I still think this is very intuitive. Remember that users generally won't be using the `batch` command or even really notice it. So it's not like they will cause this workflow with normal CLI activity. The only time this change in `save` behavior will occur is if the user pastes in a `diff` (or `dump`) and then we're giving a courtesy "heads-up" that something was wrong and they should probably look through the output because it would have scrolled off or been cleared by the save. When the user is entering commands on their own the error should be visible and they can fix as appropriate.

The other argument is what about if the user pastes in a block of commands that didn't come from a `diff` and thus didn't get the `batch` commands added? Well nothing will change in that case. The original solution addressed that by having the Configurator detect the "pasting" event and auto-wrapping the block but we decided not to go that direction.

mikeller replied:
I think that the approach to have modifiers on commands to have them behave differently is well understood by users, just like `diff` vs. `diff all`.

One more reason for not wanting to use identical commands, even more so in the case of copy / pasting backups is that, depending on whether the user pastes the backup with or without the `save` at the end, manually entering save will either save, or force save, if identical commands are used for both of these.

mikeller commented on src/main/cli/cli.c:
```diff
@@ -3745,10 +3756,54 @@ static void cliDumpRateProfile(uint8_t rateProfileIndex, uint8_t dumpMask)
 
+#ifdef USE_CLI_BATCH
+    if (commandBatchActive && commandBatchError) {
+        cliPrintCommandBatchWarning("Enter the 'save' command again to continue");
```
Probably better to not tell the user to "enter 'save' again" - this makes it sound like we do not expect them to fix the errors first. Maybe 'please fix errors, then save'?

mikeller commented on src/main/cli/cli.c:
```diff
@@ -3745,10 +3756,54 @@ static void cliDumpRateProfile(uint8_t rateProfileIndex, uint8_t dumpMask)
+#ifdef USE_CLI_BATCH
+static void cliPrintCommandBatchWarning(const char *warning)
+{
+    cliPrintErrorLinef("Errors were detected - please review before continuing");
```
Pretty much all other errors are all caps - maybe make these too?

etracer65 commented:
Rebased and changed the error messages to upper case as suggested.

I'd also like to request that we resurrect https://github.com/betaflight/betaflight-configurator/pull/1193 as it makes the handling of pasted blocks of commands transparent from the user perspective and will work with any block - not just from `diff` or `dump`. I tested it with the firmware changes from this PR and it still works properly. Needs a rebase but no other changes.

etracer65 commented:
`MIDELICF3` target overflowed.

mikeller commented:
> MIDELICF3 target overflowed.

#7735 will take care of that.
